packages/lib-runtime/src/store/coderefs.ts

```diff
@@ -0,0 +1,150 @@
+import type { AnyObject } from '@monorepo/common';
+import { CustomError, visitDeep } from '@monorepo/common';
+import * as _ from 'lodash-es';
+import type { EncodedCodeRef, CodeRef, LoadedExtension } from '../types/extension';
+import type { PluginEntryModule } from '../types/runtime';
+
+class CodeRefError extends CustomError {
+  constructor(message: string, readonly cause?: unknown) {
+    super(message);
+  }
+}
+
+const codeRefSymbol = Symbol('CodeRef');
+
+/**
+ * Check if the given object has the {@link EncodedCodeRef} shape.
+ */
+const isEncodedCodeRef = (obj: unknown): obj is EncodedCodeRef =>
+  _.isPlainObject(obj) &&
+  _.isEqual(Object.getOwnPropertyNames(obj), ['$codeRef']) &&
+  typeof (obj as EncodedCodeRef).$codeRef === 'string';
+
+/**
+ * Check if the given object is a {@link CodeRef} function.
+ */
+const isCodeRef = (obj: unknown): obj is CodeRef =>
+  typeof obj === 'function' &&
+  _.isEqual(Object.getOwnPropertySymbols(obj), [codeRefSymbol]) &&
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (obj as any)[codeRefSymbol] === true;
+
+/**
+ * Parse data from the {@link EncodedCodeRef} object.
+ *
+ * Returns `undefined` if the `$codeRef` value is malformed.
+ */
+const parseEncodedCodeRef = (
+  ref: EncodedCodeRef,
+): [moduleName: string, exportName: string] | undefined => {
+  const match = ref.$codeRef.match(/^([^.]+)(?:\.([^.]+)){0,1}$/);
+
+  if (!match) {
+    return undefined;
+  }
+
+  const moduleName = match[1];
+  const exportName = match[2] || 'default';
+
+  if (!moduleName || !exportName) {
+    return undefined;
+  }
+
+  if (moduleName !== moduleName.trim() || exportName !== exportName.trim()) {
+    return undefined;
+  }
+
+  return [moduleName, exportName];
+};
+
+/**
+ * Create a {@link CodeRef} function from the {@link EncodedCodeRef} object.
+ *
+ * The referenced object is resolved just once. Multiple {@link CodeRef} Promise resolutions
+ * yield the exact same value.
+ */
+const createCodeRef = (
+  encodedRef: EncodedCodeRef,
+  entryModule: PluginEntryModule,
+  formatErrorMessage: (message: string) => string = _.identity,
+): CodeRef => {
+  let referencedObject: unknown;
+  let referencedObjectResolved = false;
+
+  const codeRef: CodeRef = async () => {
+    if (referencedObjectResolved) {
+      return referencedObject;
+    }
```

Comment:
If this function is called twice before the first one completes, it will be processed twice?
It doesn't seem like there's a downside but the comment suggests that resolution is done only once.

Reply:
The only async bit of this function is
```ts
const moduleFactory = await entryModule.get(moduleName);
```

so if this function gets called twice in a row, it is technically possible for the 2nd execution to get ahead of the 1st execution. In any case, `moduleFactory` would be computed twice, yes. I don't think there's a solution for this, given the async nature of `moduleFactory` computation.

Reply:
> If this function is called twice before the first one completes, it will be processed twice?

Basically yes, due to `moduleFactory` async computation being part of the resolution logic.

Reply:
> It doesn't seem like there's a downside but the comment suggests that resolution is done only once.

Good point, will update the comment.

Reply:
`codeRef` function could create a shared `Promise` instance and the first call is responsible for resolving that `Promise`.

Reply:
Yup, that's exactly what I was thinking of yesterday as well, will do the change.